File: app/services/insights/adapters/sqlite.rb
module Insights::Adapters
  class SQLite < SQL
    def allowed_date_truncations
      %w(hour day month year)
    end

    def truncate_date(sql, truncation)
      raise "Bad date truncation '#{truncation}'" unless truncation.to_s.in?(allowed_date_truncations)

      sql = super(sql)
      if truncation.to_s == 'day'
        "date(#{sql})"
      else
        "datetime(#{sql}, 'start of #{truncation}')"
      end
    end

    def filter_equals(sql, type, string)
      if type == 'boolean'
        "(#{sql}) = #{quote(string == 'true' ? 't' : 'f')}"
      else
        "(#{sql}) = #{quote(string)}"
      end
    end
  end
end

File: packages/insights-api/src/insights/adapter/sql/sqlite.ts
import { ColumnType, TruncationType } from '../../definitions.d'

import SQL from './index'

export default class SQLite extends SQL {
  allowedDateTruncations () {
    return ['hour', 'day', 'month', 'year']
  }

  truncateDate (sql: string, truncation: TruncationType) {
    if (!this.allowedDateTruncations().includes(truncation)) {
      throw new Error(`Bad date truncation '${truncation}'`)
    }

    const dateSql = super.truncateDate(sql, truncation)

    if (truncation === 'day') {
      return `date(${dateSql})`
    } else {
      return `datetime(${dateSql}, 'start of ${truncation}')`
    }
  }

  filterEquals (sql: string, type: ColumnType, string: string) {
    if (type === 'boolean') {
      return `(${sql}) = ${this.quote(string === 'true' ? 't' : 'f')}`
    } else {
      return `(${sql}) = ${this.quote(string)}`
    }
  }
}
